File: app/src/main/java/com/shopcart/data/Repository.java
package com.shopcart.data;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;
import com.shopcart.data.models.Banner;
import com.shopcart.data.models.Category;
import com.shopcart.data.models.Product;
import com.shopcart.data.models.User;

import java.util.ArrayList;

public class Repository {

    private static User userData = null;
    private static ArrayList<Banner> banners = null;
    private static ArrayList<Category> categories = null;
    private static ArrayList<Product> featuredProducts = null;
    private static ArrayList<Product> bestSellProducts = null;
    private static ArrayList<Product> favouriteProducts = null;

    private Repository() {
    }

    public static boolean isDataDownloaded() {
        return banners != null && categories != null && featuredProducts != null && bestSellProducts != null && favouriteProducts != null;
    }

    public static void getData() {
        getUserData();
        getBanners();
        getCategories();
        getFeaturedProducts();
        getBestSellProducts();
        getFavouriteProducts();
    }

    public static User getUserData() {
        if (userData == null) {
            FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
            FirebaseFirestore firebaseFirestore = FirebaseFirestore.getInstance();

            if (firebaseUser != null) {
                firebaseFirestore.collection("users").document(firebaseUser.getUid()).get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                    @Override
                    public void onSuccess(DocumentSnapshot documentSnapshot) {
                        userData = documentSnapshot.toObject(User.class);
                    }
                });
            }
        }
        return userData;
    }

    public static ArrayList<Banner> getBanners() {
        if (banners == null) {
            FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
            FirebaseFirestore firebaseFirestore = FirebaseFirestore.getInstance();

            if (firebaseUser != null) {
                firebaseFirestore.collection("banners").get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                        banners = (ArrayList<Banner>) queryDocumentSnapshots.toObjects(Banner.class);
                    }
                });
            }
        }
        return banners;
    }


    public static ArrayList<Category> getCategories() {
        if (categories == null) {
            FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
            FirebaseFirestore firebaseFirestore = FirebaseFirestore.getInstance();
            if (firebaseUser != null) {
                firebaseFirestore.collection("Categories").get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                        categories = (ArrayList<Category>) queryDocumentSnapshots.toObjects(Category.class);
                    }
                });
            }
        }
        return categories;
    }

    public static ArrayList<Product> getFeaturedProducts() {
        if (featuredProducts == null) {
            FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
            FirebaseFirestore firebaseFirestore = FirebaseFirestore.getInstance();
            if (firebaseUser != null) {
                firebaseFirestore.collection("products").get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                        featuredProducts = (ArrayList<Product>) queryDocumentSnapshots.toObjects(Product.class);
                    }
                });
            }
        }
        return featuredProducts;
    }

    public static ArrayList<Product> getBestSellProducts() {
        if (featuredProducts == null) {
            FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
            FirebaseFirestore firebaseFirestore = FirebaseFirestore.getInstance();
            if (firebaseUser != null) {
                firebaseFirestore.collection("products").get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                        bestSellProducts = (ArrayList<Product>) queryDocumentSnapshots.toObjects(Product.class);
                    }
                });
            }
        }
        return bestSellProducts;
    }

    public static ArrayList<Product> getFavouriteProducts() {
        if (favouriteProducts == null) {
            FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
            FirebaseFirestore firebaseFirestore = FirebaseFirestore.getInstance();
            if (firebaseUser != null) {
                firebaseFirestore.collection("products").get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                        favouriteProducts = (ArrayList<Product>) queryDocumentSnapshots.toObjects(Product.class);
                    }
                });
            }
        }
        return favouriteProducts;
    }
}

File: app/src/main/java/com/shopcart/data/Repository.kt
package com.shopcart.data

import android.content.SharedPreferences
import android.content.res.Resources
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ktx.toObject
import com.google.firebase.firestore.ktx.toObjects
import com.shopcart.R
import com.shopcart.data.models.Banner
import com.shopcart.data.models.Category
import com.shopcart.data.models.Product
import com.shopcart.data.models.User
import com.shopcart.utilities.Constants
import com.shopcart.utilities.Resource
import com.shopcart.utilities.safeCall
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import javax.inject.Inject

class Repository @Inject constructor(
    private val sharedPreferences: SharedPreferences,
    private val firebaseAuth: FirebaseAuth,
    private val fireStore: FirebaseFirestore
) {
    var onBoardingState: Boolean = false
        get() {
            field = sharedPreferences.getBoolean(Constants.ON_BOARDING_STATE_KEY, false)
            return field
        }
        set(value) {
            field = value
            sharedPreferences.edit()
                .putBoolean(Constants.ON_BOARDING_STATE_KEY, value).apply()
        }

    suspend fun getProfile(): Resource<User> {
        return withContext(Dispatchers.IO) {
            safeCall {
                val result =
                    fireStore.collection("users").document(firebaseAuth.currentUser!!.uid).get()
                        .await().toObject<User>() as User


                Resource.Success(result)
            }
        }
    }

    suspend fun signUp(email: String, password: String): Resource<AuthResult> {
        return withContext(Dispatchers.IO) {
            safeCall {
                // Create new user
                val result =
                    firebaseAuth.createUserWithEmailAndPassword(email, password)
                        .await()

                // Send email verification
                result.user!!.sendEmailVerification().await()

                // Add User To Database
                result.user!!.uid.let {
                    val newUser = User(it, email)
                    fireStore.collection("users").document(it).set(newUser).await()
                }

                Resource.Success(result)
            }
        }
    }

    suspend fun signIn(email: String, password: String): Resource<AuthResult> {
        return withContext(Dispatchers.IO) {
            safeCall {
                val result = firebaseAuth.signInWithEmailAndPassword(email, password).await()
                if (result.user!!.isEmailVerified)
                    Resource.Success(result)
                else Resource.Error(
                    Resources.getSystem().getString(R.string.login_toast_please_verify)
                )
            }
        }
    }

    suspend fun signOut() {
        return withContext(Dispatchers.IO) {
            firebaseAuth.currentUser?.let { firebaseAuth.signOut() }
        }
    }

    suspend fun getBanners(): Resource<ArrayList<Banner>> {
        return withContext(Dispatchers.IO) {
            safeCall {
                val result = fireStore.collection("banners").get().await()
                    .toObjects<Banner>() as ArrayList<Banner>

                Resource.Success(result)
            }
        }
    }

    suspend fun getCategories(): Resource<ArrayList<Category>> {
        return withContext(Dispatchers.IO) {
            safeCall {
                val result = fireStore.collection("Categories").get().await()
                    .toObjects<Category>() as ArrayList<Category>

                Resource.Success(result)
            }
        }
    }

    suspend fun getFeaturedProducts(): Resource<ArrayList<Product>> {
        return withContext(Dispatchers.IO) {
            safeCall {
                val result = fireStore.collection("products")
                    .whereEqualTo("isFeatured", true)
                    .get().await()
                    .toObjects<Product>() as ArrayList<Product>

                Resource.Success(result)
            }
        }
    }

    suspend fun getBestSellProducts(): Resource<ArrayList<Product>> {
        return withContext(Dispatchers.IO) {
            safeCall {
                val result = fireStore.collection("products")
                    .whereEqualTo("isBestSell", true)
                    .get().await()
                    .toObjects<Product>() as ArrayList<Product>

                Resource.Success(result)
            }
        }
    }

    suspend fun getFavouriteProducts(): Resource<ArrayList<Product>> {
        return withContext(Dispatchers.IO) {
            safeCall {
                val result = fireStore.collection("products").get().await()
                    .toObjects<Product>() as ArrayList<Product>

                Resource.Success(result)
            }
        }
    }
}